### ISSUE TITLE
Does GM.xmlHttpRequest send cookies?

### ISSUE THREAD

**Issue by erosman** [greasemonkey/greasemonkey#2826]
ref:: FF59.0a1, Win65 GM4.3beta2

In a script that requires cookies (to stay logged in), the results of `GM.xmlHttpRequest` shows as a not-logged-in-user.

It seems that the cookies (session ones in this case) are not passed.
Is that the case?
How can I test it?
How can I fix it?

**Comment by NoneGiven** [comment]
It does not. The old `GM_xmlhttpRequest` didn't either, but it did allow you to set the `Cookie` header with whatever you wanted. I have a PR to allow  `GM.xmlHttpRequest` to set forbidden headers such as that one: #2723

**Comment by erosman** [comment]
> It does not. The old GM_xmlhttpRequest didn't either,

Not correct.. the old one did.
This is the same script that works fine with GM3 but not with GM4
XHR  by default sends all credentials, unless GM4 is using `fetch()`

**Comment by NoneGiven** [comment]
Is that so? I only ever used it for cross-origin requests, since it was unnecessary otherwise, and it would not send any cookies unless manually specified in that case.

I had another PR that would give the new version the ability to automatically use the relevant cookies (and we could limit it to same-origin requests to match the old behavior), but I did it wrong and closed it so we can focus on #2723 which would need to happen first anyway.

>XHR by default sends all credentials

There are no credentials implicitly available when it's sent from a background script.

**Comment by erosman** [comment]
> Not when it's sent from a background script.

Hm ... makes sense

In GM3 there was no such issue and yes, there is no cross-origin requests as the current page and the target belong to the same domain.

Has something changed in `GM.xmlHttpRequest` behaviour in GM4?

**Comment by Sxderp** [comment]
> Has something changed in GM.xmlHttpRequest behaviour in GM4?

GM4 sends the request from the background page. Therefore the origin is the `moz-extension://` and not the web page. As such it doesn't have easy access to the credentials of the page that is running the script.

However, if you're not doing a cross-origin XHR, then you should be able to use a standard XHR, and not have to go through `GM.xmlHttpRequest`, no?

**Comment by erosman** [comment]
> GM4 sends the request from the background page.
I thought so ... 

> if you're not doing a cross-origin XHR, then you should be able to use a standard XHR, and not have to go through GM.xmlHttpRequest, no?
Sure.... That is what I will do.
It got me baffled since a working GM3 script wasnt working as GM4

**Comment by NoneGiven** [comment]
If you want your normal XHR to have the page cookies, you'll have to get the XHR object from the page through a wrapper.

```
XPCNativeWrapper(new window.wrappedJSObject.XMLHttpRequest());
```

**Comment by erosman** [comment]
> If you want your normal XHR to have the page cookies, it won't "just work;"

I have just noticed :(

A lot of hoops to go through ...

**Comment by NoneGiven** [comment]
Pony or not, I do want to take another crack at this sometime. Simpler and much more sensible approach as compared to my last attempt:

* If the destination domain of the XHR matches the domain of the tab calling `GM.xmlHttpRequest`:
    * Use the cookie API to get the cookies for that domain and set them.
* If the user specifies their own `Cookie` header, even an empty one, skip the above step entirely.

**Comment by NoneGiven** [comment]
@arantius I was drafting the changes for the proposal above, and a possible enhancement occurred to me:

* If the user specifies their own `Cookie` header, send those values and skip the following steps.
* If the destination domain of the XHR matches the domain of the tab calling `GM.xmlHttpRequest`,
\- OR -
* If the domains don't match, but the user passes `withCredentials: true` to `GM.xmlHttpRequest`:
    * Use the cookie API to get the cookies for the XHR destination domain and send those values.

Thoughts? This would give `GM.xmlHttpRequest` cookie behavior (and security) identical to a normal XHR.

Also, just double-checking an unlikely security concern: the messaging ports used for things like `GM.xmlHttpRequest` can only be accessed by GM's own content scripts, not any old JS, right?

**Comment by NoneGiven** [comment]
@arantius I was drafting the changes for the proposal above, and a possible enhancement occurred to me:

* If the user specifies their own `Cookie` header, send those values and skip the following steps.
* If the destination domain of the XHR matches the domain of the tab calling `GM.xmlHttpRequest`,
- OR -
* If the domains don't match, but the user passes `withCredentials: true` to `GM.xmlHttpRequest`:
    * Use the cookie API to get the cookies for the XHR destination domain and send those values.

Thoughts? This would give `GM.xmlHttpRequest` cookie behavior (and security) identical to a normal XHR.

Also, just double-checking an unlikely security concern: the messaging ports used for things like `GM.xmlHttpRequest` can only be accessed by GM's own content scripts, not any old JS, right?

**Comment by arantius** [comment]
> @arantius I was drafting the changes for the proposal above, and a possible enhancement occurred to me...

This all sounds good.
